### DISCUSSION TITLE
MVP Scope Discussion

### DISCUSSION THREAD

**Discussion by muhuk** [muhuk/yer#4] [General]
### What should we have completed for a `version 0.1`  release (public release or not)?

My list of features:

- A basic editor UI.

  - With a basic menu
  - A 3D viewport
  - A sidebar (can be placeholder/empty)

- Ability to create a new project.
- A base layer that can be flat or some kind of noise.
- Ability to save and load projects.
- Ability to export as a heightmap.

  - Export size can be fixed as 1024x1024.

- Preview border and export border can be fixed as `[-1.0, -1.0] to [1.0, 1.0]`
- Set up unit scale (what does `1 unit` mean in terms of meters/kilometers) in project.

Let's discuss.

**Comment by sniperliu** [comment]
I think `v0.1` can be a flat noise generator.

 - Menus: [New, Load, Export]
 - Side Bar: based on noise & its parameters, a generate button
 - View: 2D or 3D

Is `unit` that import in a vector space? I have no experience on this.

If we want infinite space, will that be too early to decide `unit`?

**Reply by muhuk** [reply]
> flat noise generator

What do you mean by `flat noise`? Flat is one single value and noise is variable by definition.

> Side Bar: based on noise & its parameters, a generate button

There should be no generate button. By generation if you mean the noise kernel, that should be near instantaneous.

> View: 2D or 3D

I was thinking to drop 2D preview for simplicity.

> Is unit that import in a vector space?
> If we want infinite space, will that be too early to decide unit?

Unit is fundamental because we will be showing measurements in UI in this unit. Also because it is needed to calculate texel density. But mostly for UI purposes, we should never show internal units to the user. Internal unit should be 100% internal.

It is possible to work units in later, but implementing it now is not too difficult and adding it later will likely result in duplicate work.

The unit basically is a tuple of unit type and scale, stored at the project level.

```rust
enum UnitType {
  Kilometer,
  Meter
}

struct Unit {
  unit_type: UnitType,
  scale: f32
}
```

**Comment by sniperliu** [comment]
In this stage, do we need to organize the project as workspace?

```toml
[workspace]
members = [ "libs/*", ]
resolver="2"
```

```
libs/fbm
libs/ui
```

**Reply by muhuk** [reply]
We can.  But I don't know what's the best way to slice and dice things.  I mean if we do it now, I suspect it won't be optimal and we will want to restructure workspaces again in the future. I feel like postponing this is a good idea.

Having said all that I think eventually we will want to do this. I think eventually we will also want to have workspaces one of us is primarily responsible.

Also I am not a big fan of one-workspace-per-top-level-module thing.

**Reply by sniperliu** [reply]
Ok, we agree not need in v0.1.
